### PR TITLE
Fixes #21244 - Rails 5 upgrade test failures

### PR DESCRIPTION
**Pull request by lzap**
Remaining 2/3 R5 errors fixed, but there is one which wrote @dLobato and I don't understand. The assertion in the test looks incorrect to me, assert takes one parameter, should that be `assert_equal`? In any case, architecture is not being set there. Can you take a look? It's this one:

```ruby
module ForemanDiscovery
  module Concerns
    class HostsControllerTest < ActionController::TestCase
      tests ::HostsController

      context 'hosts controller requests from discovered_hosts url' do
        test 'get "host" params from "discovered_hosts" params' do
          architecture = FactoryGirl.create(:architecture)
          discovered_host_params = {
            'discovered_host' => { 'architecture_id' => architecture.id }
          }

          @request.stubs(:path).
            returns(architecture_selected_discovered_hosts_path)

          post :architecture_selected, discovered_host_params, set_session_user
          assert assigns('architecture'), architecture # THIS FAILS, 1st param is nil
        end
      end
    end
  end
end
```

### CODE REVIEW

File: app/controllers/api/v2/discovered_hosts_controller.rb
Comment (lzap) [inline app/controllers/api/v2/discovered_hosts_controller.rb:101]: This is not a mass assignement, so it is safe to do. We do the same in core:

app/controllers/api/v2/hosts_controller.rb
289:        state = @host.import_facts(params[:facts].to_unsafe_h, detected_proxy)

File: test/functional/api/v2/discovery_rules_controller_test.rb
Comment (lzap) [inline test/functional/api/v2/discovery_rules_controller_test.rb:41]: Looks like ID parser in Rails 5 did not catch this anymore, it was a typo anyway. Integer is expected here.

All the rest was just cleaning up Ruby compiler warnings - missing parentheses and unused variable.

**Review by dLobatog — COMMENTED** [review]
@lzap It should be either:

- To make sure `@architecture` is set to something (not nil) after the controller call.

```ruby
assert assigns('architecture')
```

To make sure `@architecture` and `architecture` are the same:

```ruby
assert_equal architecture, assigns('architecture')  
```

**Comment by lzap** [comment]
Right, that's what I though. Any clue why `assigns('architecture')` is `nil` with Rails 5? I can't figure it out.

**Comment by dLobatog** [comment]
@lzap I think here's the culprit: 

https://github.com/rails/rails/issues/18950

If I'm not wrong you can check `assert_equal architecture, @architecture`

**Comment by lzap** [comment]
Unfortunately, the variable is really nil:

--- expected
+++ actual
@@ -1 +1 @@
-#<Architecture id: 501905020, name: "x86_64-27", created_at: "2017-10-09 11:46:46", updated_at: "2017-10-09 11:46:46">
+nil

**Comment by lzap** [comment]
Finally solved it, spend two days digging in core but this was all about controller attributes not available in unit tests anymore. You can still access it via `@controller` tho. Tests are green for both R4 and R5.

File: test/functional/foreman_discovery/concerns/hosts_controller_extensions_test.rb
Comment (ares) [inline test/functional/foreman_discovery/concerns/hosts_controller_extensions_test.rb]: in Rails 5 they made functional tests equal to integration tests, we should perhaps test that response contains architecture.name or something and what was the response code

**Review by ares — CHANGES_REQUESTED** [review]
one nit, otherwise looks good to me, thanks for inline comments, it saved me some time

File: test/functional/foreman_discovery/concerns/hosts_controller_extensions_test.rb
Comment (lzap) [reply test/functional/foreman_discovery/concerns/hosts_controller_extensions_test.rb]: Agree but see the comment - the response actually does not contain architecture but operating systems associated with this architecture. It's confusing. I could however do some factorygirl magic to create some associations and test that if you insist.

Comment (lzap) [reply test/functional/foreman_discovery/concerns/hosts_controller_extensions_test.rb]: Ok changed it to integration test.

**Comment by lzap** [comment]
Anything else?

**Comment by ares** [comment]
LGTM, I haven't tested it, if you could trigger the jenkins job manually to confirm, that would be great (or let me know if you don't know how). I don't have merge permissions but after jenkins confirms I think this can be merged.

**Comment by lzap** [comment]
I think I have permissions but how do I do that? Which job is that?

**Comment by lzap** [comment]
Lemme try [test]

**Comment by timogoebel** [comment]
This looks good. If tests pass on Jenkins ([test] does not seem to help, but I don't know any other way as well), I'll happily merge this.

**Comment by ares** [comment]
test won't help until develop branch is green, you need to trigger test_plugin_pull_request, triggered at http://ci.theforeman.org/job/test_plugin_pull_request/4210/

**Comment by ares** [comment]
the result is green, this can (and should be merged)

**Comment by lzap** [comment]
http://ci.theforeman.org/job/test_plugin_pull_request/4210/ ALL GREEN I am skipping our rule and merging this time.

**Pull request merged**
